### DISCUSSION TITLE
Нужна ли нам куча билдеров у EntityStateBuilder?

### DISCUSSION THREAD

**Discussion by IlyaMax** [surfstudio/SurfGear#213] [Ideas]
На данный момент у нас есть уже три билдера и планируется добавить еще один. Вот список:
```
typedef LoadingWidgetBuilder<T> = Widget Function(BuildContext context, T? data);

typedef DataWidgetBuilder<T> = Widget Function(BuildContext context, T data);

typedef ErrorWidgetBuilder = Widget Function(BuildContext context, Exception? e);
// Этот билдер хотим добавить в 3.0
typedef DataErrorWidgetBuilder<T> = Widget Function(BuildContext context, T? data, Exception? e);
```

Какие есть с этим проблемы:

1. Не совсем очевидно что если я укажу loadingBuilder и loadingChild второй у меня никогда не вызовется. Аналогично по error. 
2. У билдеров есть приоритеты, даже если в доке указать приоритеты неочевидно что используется.

Эту проблему не решает декомпозиция абсолютно, т.к. всё-равно нужно провайдить ключ isError или isLoading. В итоге код обрастает кучей ненужных строк, когда можно было бы проще.

А как проще? Оставить только loadingChild и errorChild, contentBuilder(он же builder, он же child).  Если нужно какое-то кастомное условие то юзать StreamedStateBuilder<EntityState<T>>.

**Comment by pretorean** [comment]
ну тогда может быть убрать child и оставить только билдер

**Reply by IlyaMax** [reply]
можно и так, но что делать с DataErrorWidgetBuilder?

**Reply by saturov** [reply]
У меня сразу после прочитанного возникла такая же мысль, что нужно убрать `child`, оставив только `builder`.

**Comment by saturov** [comment]
Зачем нужен этот `DataErrorWidgetBuilder`? Если так нужны данные, добавить их в имеющийся `ErrorWidgetBuilder`
